🚀 Pull Request: Unified Status Handling, Owner Relations, and Full Booking Integration

Miku21750 opened the pull request:
# 🚀 Pull Request: Unified Status Handling, Owner Relations, and Full Booking Integration

## ✅ Summary

This PR includes major schema, API, and UI updates across the application to improve data consistency, user context integration, and provide flexible entity closure without relying solely on bookings.

---

## 🔧 Schema Changes

### 🧱 Status Enums and Owner Relations
- Converted the following status fields to use strict enums:
  - `CaseStatus`
  - `OrderStatus`
  - `SystemStatus`
  - `MaterialOrderLineItemStatus`
- Made enums required with default values.
- Replaced `Owner` (string) in `workorder` and `materialorder` with `OwnerID` (foreign key to `User`).
- Added indexes and foreign key constraints for `OwnerID`.

### 🗃️ Booking Schema
- Created tables:
  - `Resource`, `ResourceAccount`, `SubkTechnician`, `Bookings`, `BookingDetails`
- Changed primary key types to `VARCHAR` for flexibility.
- Removed `SubkTechnicianLearner` table and cleaned related references.
- Linked `Bookings` to `WorkOrder` via `WOID`, and to `User` via `CreatedBy`.

---

## 🚀 API Enhancements

- Added `PATCH` endpoints for:
  - `MaterialOrder`
  - `MaterialOrderLineItems`
  - `Bookings`
- Added full `POST` and `GET` support for bookings:
  - `GET /api/bookings`, `/resources`, `/resourceAccounts`, `/subkTechnicians`
  - `POST /api/bookings/create`
  - `PATCH /api/bookings/[BookingId]`
- Integrated `OwnerID` from JWT into creation logic.
- Refactored handlers for better parameter consistency.

---

## 💡 Frontend Features

### 📆 Booking Management
- Dynamic `ServiceBooking` component with props (`BookingId`, `woid`) and editable state.
- `<NewBookableResourceBooking>` component enables seamless creation and redirect to booking detail.
- Added bookings list in `WorkOrder` detail view.
- Implemented live search with debounce for Resource, Account, and Technician fields.
- Integrated `CreatedBy` via user context.

### 📝 Save & Close Functionality
- Implemented "Save & Close" for:
  - Case → `CaseStatus: 'Close'`
  - Work Order → `SystemStatus: 'CLOSED_POSTED'`
  - Material Order → `OrderStatus: 'Closed'`
  - Line Item → `Status: 'Closed'`
- SweetAlert used for confirmation and success feedback.
- Navigation and UI updates post-close.

### 🚫 Read-Only Indicators
- Added yellow alert banners on closed items (Case, Work Order, Material Order, Line Item).
- Tabs and forms disabled when status is closed.
- Data passed to components for conditional UI rendering.

---

## 🛠️ Known Limitation & Suggestion

**Problem**: Bookings are not always completed due to input/data bugs, making it difficult to close higher-level records.

**Proposed Solution**: Allow status-only closure without booking dependency.

### 🔧 Schema Adjustment (if needed)
- Ensure nullable `Status` field exists on:
  - `Case`
  - `WorkOrder`
  - `MaterialOrder`
  - `MaterialOrderLineItems`

### 🔁 Utility Function
```ts
async function closeEntityStatus(entityType: "case" | "workorder" | "materialorder" | "lineitem", id: string) {
  // PATCH status to "Closed"
}
```

### 💡 Integration Points
- Add "Close Case" button for each detail screen.
- Allow manual closure using `closeEntityStatus`.
- Show confirmation modal before closure.

---

## 📌 To Do (Post-Merge)
- [x] Verify presence of `Status` fields across all affected tables.
- [ ] Finalize booking form bugs and complete booking lifecycle.
- [ ] Continue UI improvements and error validations.
- [ ] Confirm new columns exist in all required tables
- [x] Implement PATCH logic for each entity's status
- [x] Add "Close" buttons to each detail screen
- [ ] Optional: Add confirmation modal (Swal) before closing

---

## 🧪 Testing Notes
- Test full create/edit/close flow for:
  - Case
  - Work Order
  - Material Order
  - Line Items
  - Bookings
- Test auto-fill via user context and confirm all PATCH routes update status correctly.

---

## 🧾 Related Issues
- Implements #booking-schema
- Closes #close-by-status
- Addresses #owner-refactor

---

Please review and provide feedback or approval to proceed with deployment.

Miku21750 committed:
Merge branch 'pr/51' into feature-miku21-before-merge-langit

Miku21750 committed:
Merge branch 'pr/51' into feature-miku21-merging

Miku21750 committed:
Merge branch 'pr/51' into feature-miku21-merging

Miku21750 committed:
feat(workflow): allow case and related records to be closed without booking completion

## feat(workflow): allow case and related records to be closed without booking completion

### ✅ Summary of Changes

#### 🏗️ Database Schema Updates
- Added `Status` column (if not present) to the following tables:
  - `Case`
  - `WorkOrder`
  - `MaterialOrder`
  - `MaterialOrderLineItems`

#### 🔄 Status Management Logic
- Introduced a utility function to patch status to `closed` across related records
- Ensured foreign key relationships do not block record closure when bookings are incomplete
- Made the status update logic reusable and modular

#### 🚀 New Functional Flow
- Implemented ability to mark a case and all associated records as `closed` regardless of booking state
- Added handler/function to:
  - Validate closure conditions (if any minimal logic is required)
  - Patch statuses using API calls
  - Be reused from case detail or work order detail view components

### 🧠 Rationale
- Enables users to progress and close out cases in exceptional scenarios (e.g., incomplete bookings)
- Prevents workflow bottlenecks caused by data-read failures in `Bookings`

### 🧹 Improvements from Previous Booking Flow
- Made `ServiceBooking` component dynamic with props: `BookingId` and `woid`
- Debounced and optimized input search in resource selection
- Improved dropdown cleanup on empty input or tab navigation
- Added validation and correct icon usage for `Swal.fire` messages
- Displayed related bookings in work order detail view
- Replaced static "New Bookable Resource" with `<NewBookableResourceBooking />` component
- Included `createdBy` using JWT user context
- Enhanced pagination/search for resource-related API endpoints

---

### ❓Next Suggestion
Would you like to add a **"Delete Booking"** button with confirmation modal in `ServiceBooking`?

Miku21750 committed:
Merge branch 'feature' into feature-miku21-merging

Mackovlak commented on test-vite/src/main.jsx:
```diff
@@ -37,7 +37,9 @@ createRoot(document.getElementById('root')).render(
             <Route path='/material-order/:moid' element={<MaterialOrder />}/>
             <Route path='/mo_detail/:molineid' element={<MoDetail />}/>
-            <Route path='/bookings' element={<Bookings/>}/>
+            <Route path='/bookings' element={<Bookings />} />]
```
] ????

Mackovlak reviewed (approved):
## Review Report #53 
## Added
* New migration file for booking
```
npx prisma generate
```
for update new database
*  Adding Route for booking and bookingId
*  Adding several improvement for workorder route
*  Adding new package in test-vite directory
```
npm i 
```
for installing new package

* Adding different approach for service and booking component

## Seems off
* Commented code
* Need bug fixes  for merging status

# Merging failure 
## Need more Info about the bug

Miku21750 committed:
refactor(schema+api+ui): unify status enums, add owner relations, and implement close actions

refactor(schema+api+ui): unify status enums, add owner relations, and implement close actions

### 🔧 Schema Changes
- Refactored `CaseStatus`, `OrderStatus`, `SystemStatus`, and `MaterialOrderLineItemStatus` fields to use enums for stricter validation
- Made status fields non-nullable with appropriate default values
- Replaced `Owner` (string) in `materialorder` and `workorder` with `OwnerID` (foreign key to `User`)
- Added indexes and foreign key constraints for `OwnerID`

### 🚀 API Enhancements
- Added `PATCH` endpoints for Material Orders and Line Items
- Updated create-order logic to use `OwnerID` from JWT user context
- Standardized enum value usage (`CLOSED_POSTED`, `Closed`, etc.) in request handlers
- Refactored route handlers to use destructured params for clarity and consistency

### ✅ UX Features: Save & Close Functionality
- Added "Save & Close" buttons to:
  - Case → updates `CaseStatus` to `'Close'`
  - Work Order → updates `SystemStatus` to `'CLOSED_POSTED'`
  - Material Order → updates `OrderStatus` to `'Closed'`
  - Line Item → updates `Status` to `'Closed'`
- Integrated SweetAlert confirmation modals and success feedback
- Implemented navigation and form locking logic post-closure

### 🛡️ Read-Only UI Indicators
- Displayed yellow banners for closed:
  - Work Orders (when `SystemStatus === 'CLOSED_POSTED'`)
  - Material Orders (`OrderStatus === 'Closed'`)
  - Line Items (`Status === 'Closed'`)
- Disabled editing on closed records and updated UI layout
- Passed `workOrders` and `materialOrders` as props for accurate tab rendering

---
This commit finalizes end-to-end support for managing ownership, status lifecycles, and UI handling of closed entities across the Case → Work Order → Material Order → Line Item hierarchy.

Miku21750 committed:
Merge branch 'feature' into feature-miku21-merging

Mackovlak reviewed (approved):
# Review Report #53 
## After Rebuilt Feature
* Fix multiple issue in case close
* Adding booking operated booking system
* integrated save feature to detail

Miku21750 commented:
it still a bug thou
